addons/bus/websocket.py

```diff
@@ -0,0 +1,883 @@
+_XOR_TABLE = [bytes(a ^ b for a in range(256)) for b in range(256)]
+
+
+@functools.total_ordering
+class Frame:
+    # This class implements the `__lt__` method in order for frames to
+    # be stored in a `PriorityQueue`: ping/pong frames are prioritary.
+    _frames_sent = count(0)
+
+    def __init__(
+        self,
+        opcode,
+        payload=b'',
+        fin=True,
+        rsv1=False,
+        rsv2=False,
+        rsv3=False
+    ):
+        self._send_order = next(self._frames_sent)
+        self.opcode = opcode
+        self.payload = payload
+        self.fin = fin
+        self.rsv1 = rsv1
+        self.rsv2 = rsv2
+        self.rsv3 = rsv3
+
+    def __eq__(self, other):
+        if not isinstance(other, Frame):
+            return False
+        return other.opcode == self.opcode and self.payload == other.payload
+
+    def __lt__(self, other):
+        if not isinstance(other, Frame):
+            raise TypeError("< only supported between instances of 'Frame'")
+        if (
+            self.opcode in HEARTBEAT_OP and
+            other.opcode in HEARTBEAT_OP or
+            self.opcode in DATA_OP and other.opcode in DATA_OP
+        ):
+            return self._send_order < other._send_order
+        return self.opcode in HEARTBEAT_OP
+
+
+class CloseFrame(Frame):
+    def __init__(self, code, reason):
+        if code not in VALID_CLOSE_CODES and code not in RESERVED_CLOSE_CODES:
+            raise InvalidCloseCodeException(code)
+        payload = struct.pack('!H', code)
+        if reason:
+            payload += reason.encode('utf-8')
+        self.code = code
+        self.reason = reason
+        super().__init__(Opcode.CLOSE, payload)
+
+
+class Websocket:
+    _instances = set()
+    _event_callbacks = defaultdict(set)
+    # Maximum size for a message in bytes, whether it is sent as one
+    # frame or many fragmented ones.
+    MESSAGE_MAX_SIZE = 2 ** 20
+    # Proxies usually close a connection after 1 minute of inactivity.
+    # Therefore, a PING frame have to be sent if no frame is either sent
+    # or received within CONNECTION_TIMEOUT - 15 seconds.
+    CONNECTION_TIMEOUT = 60
+    INACTIVITY_TIMEOUT = CONNECTION_TIMEOUT - 15
+    # How many requests can be made in excess of the given rate.
+    RL_BURST = int(config['websocket_rate_limit_burst'])
+    # How many seconds between each request.
+    RL_DELAY = float(config['websocket_rate_limit_delay'])
+
+    def __init__(self, socket, session):
+        # Session linked to the current websocket connection.
+        self._session = session
+        self._socket = socket
+        self._close_sent = False
+        self._outgoing_frame_queue = PollablePriorityQueue()
+        self._timeout_manager = TimeoutManager()
+        # Used for rate limiting.
+        self._incoming_frame_timestamps = deque(maxlen=type(self).RL_BURST)
+        self.state = ConnectionState.OPEN
+        type(self)._instances.add(self)
```

Comment:
We use this artifact in order to close all websocket upon ctrl-c. See Do you have any alternative suggestion so that we can gracefully close all the websocket upon ctrl-c?

See `Websocket._on_server_stop`